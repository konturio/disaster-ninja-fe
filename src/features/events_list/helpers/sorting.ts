import type { Event } from '~core/types';

export function sortEventsByStartDate(events: Event[], order: 'asc' | 'desc'): Event[] {
  return [...events].sort((a, b) => {
    const dateA = new Date(a.startedAt ?? 0).getTime();
    const dateB = new Date(b.startedAt ?? 0).getTime();
    return order === 'desc' ? dateB - dateA : dateA - dateB;
  });
}
