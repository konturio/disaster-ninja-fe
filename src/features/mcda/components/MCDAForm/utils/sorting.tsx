import type { SelectableItem } from '@konturio/ui-kit';

export function sortByAlphabet(items: SelectableItem[]) {
  return items.sort((axis1, axis2) =>
    axis1.title?.localeCompare(axis2.title, undefined, { sensitivity: 'base' }),
  );
}

/**
 * Sorts an array in place. Item moves higher if the substring occurs at the start
 * of a word boundary. The earlier the occurence - the higher the item moves.
 */
export function sortByWordOccurence(items: SelectableItem[], substring: string) {
  items.sort((axis1, axis2) => {
    const regex = new RegExp(`\\b(${substring})`, 'i');
    const NOT_FOUND = 10000;
    const axis1FirstWordIndex = axis1.title.search(regex);
    const axis2FirstWordIndex = axis2.title.search(regex);
    return (
      (axis1FirstWordIndex >= 0 ? axis1FirstWordIndex : NOT_FOUND) -
      (axis2FirstWordIndex >= 0 ? axis2FirstWordIndex : NOT_FOUND)
    );
  });
}
