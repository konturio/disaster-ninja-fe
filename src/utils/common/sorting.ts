import type { SelectableItem } from '@konturio/ui-kit';

export const sortByKey =
  <T extends Record<string, unknown>>(key: string, direction: 'asc' | 'desc') =>
  (a: T, b: T) => {
    const aVal = a[key];
    const bVal = b[key];
    if (typeof aVal === 'number' && typeof bVal === 'number') {
      return direction === 'desc' ? bVal - aVal : aVal - bVal;
    }
    return 0;
  };

/**
 * Sorts an array in place alphabetically.
 **/
export function sortByAlphabet<T>(items: T[], getValue: (item: T) => string): T[] {
  return items.sort((item1, item2) =>
    getValue(item1).localeCompare(getValue(item2), undefined, { sensitivity: 'base' }),
  );
}

/**
 * Sorts an array in place. Item moves higher if the substring occurs at the start
 * of a word boundary. The earlier the occurrence - the higher the item moves.
 */
export function sortByWordOccurrence<T>(
  items: T[],
  getItemValue: (item: T) => string,
  word: string,
): T[] {
  return items.sort((item1, item2) => {
    const regex = new RegExp(`\\b(${word})`, 'i');
    const NOT_FOUND = 10000;
    const axis1FirstWordIndex = getItemValue(item1).search(regex);
    const axis2FirstWordIndex = getItemValue(item2).search(regex);
    return (
      (axis1FirstWordIndex >= 0 ? axis1FirstWordIndex : NOT_FOUND) -
      (axis2FirstWordIndex >= 0 ? axis2FirstWordIndex : NOT_FOUND)
    );
  });
}
