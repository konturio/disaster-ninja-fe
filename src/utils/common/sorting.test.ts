import { expect, describe, it } from 'vitest';
import { sortByAlphabet, sortByWordOccurrence } from './sorting';
import type { SelectableItem } from '@konturio/ui-kit';

describe('sortByWordOccurrence', () => {
  it('should sort items based on earlier occurrence of the substring at a starting word boundary', () => {
    const items: SelectableItem[] = [
      { title: 'CCC. Lower case word', value: '1' },
      { title: 'BB. Words can be plural', value: '2' },
      { title: 'A. Word', value: '3' },
      { title: 'Nonword. Not at the start boundary, does not count', value: '4' },
      { title: 'Word at the start of the line', value: '5' },
    ];
    sortByWordOccurrence(items, (item) => item.title, 'word');
    expect(items).toMatchSnapshot();
  });
});

describe('sortByAlphabet', () => {
  it('should sort items alphabetically, ignoring case', () => {
    const items: SelectableItem[] = [
      { title: 'CCC', value: '1' },
      { title: 'BB', value: '2' },
      { title: 'Ab', value: '3' },
      { title: 'ac longer item', value: '4' },
      { title: '999 numbers are earlier than letters', value: '5' },
    ];
    sortByAlphabet(items, (item) => item.title);
    expect(items).toMatchSnapshot();
  });
});
