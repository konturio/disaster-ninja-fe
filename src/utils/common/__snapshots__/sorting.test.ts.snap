// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sortByAlphabet > should sort items alphabetically, ignoring case 1`] = `
[
  {
    "title": "999 numbers are earlier than letters",
    "value": "5",
  },
  {
    "title": "Ab",
    "value": "3",
  },
  {
    "title": "ac longer item",
    "value": "4",
  },
  {
    "title": "BB",
    "value": "2",
  },
  {
    "title": "CCC",
    "value": "1",
  },
]
`;

exports[`sortByWordOccurence > should sort items based on earlier occurence of the substring at a starting word boundary 1`] = `
[
  {
    "title": "Word at the start of the line",
    "value": "5",
  },
  {
    "title": "A. Word",
    "value": "3",
  },
  {
    "title": "BB. Words can be plural",
    "value": "2",
  },
  {
    "title": "CCC. Lower case word",
    "value": "1",
  },
  {
    "title": "Nonword. Not at the start boundary, does not count",
    "value": "4",
  },
]
`;
